modules/lang-painless/src/test/java/org/elasticsearch/painless/ElvisTests.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.painless;
+
+import static java.util.Collections.singletonMap;
+import static org.hamcrest.Matchers.greaterThan;
+import static org.hamcrest.Matchers.lessThan;
+
+/**
+ * Tests for the Elvis operator ({@code ?:}).
+ */
+public class ElvisTests extends ScriptTestCase {
+    public void testBasics() {
```

Comment:
Yeah, to get a primitive at this point you have cast. Like
```
org.elasticsearch.common.geo.GeoPoint p = (org.elasticsearch.common.geo.GeoPoint) doc.somepoint.value;
double lat = p?.lat ?: 0;
```
would do it, I believe. I expect there are other ways but they are much less intuitive.

I'm onboard with disallowing the lhs of `?:` to be a primitive. That is in this proposal actually. 

I feel like disallowing the return value of `?.` from being a primitive would push people to the def case when they don't have to be there. And, yeah, it is all pretty moot at this point because fields are always coming back from ES as defs and you have to cast anyway. We'll get to the point where we can specialize the types of those, I think, eventually. It'll be a while though, I expect.

Could we get away with saying "`?.` boxes if needed and `?:` unboxes if needed"? Those just feel so natural to me, so right in the context of each operator. And they mesh so well when used together, even begging for that `x?.primitiveField ?: primitiveDefault` optimization we'd talked about.

I suppose part of my frustration is that this whole no unboxing thing just sort of jumped out at me. I'd never really realized that painless behaved in this fairly unexpected way. For the most part it doesn't come up but when it does it runs so counter to the way I expect something to behave on the JVM I haven't wrapped my head around it.

I'm fine with keeping it as it is, or, rather, I admit that reworking it is not worth considering in the short term. I'm certainly open to keeping it forever if we have a solid performance argument around it that I can internalize so I can explain it.